drivers/ethernet/eth_stm32_hal.c

```diff
@@ -241,16 +242,40 @@ static void rx_thread(void *arg1, void *unused1, void *unused2)
 	__ASSERT_NO_MSG(dev_data != NULL);
 
 	while (1) {
-		k_sem_take(&dev_data->rx_int_sem, K_FOREVER);
-
-		while ((pkt = eth_rx(dev)) != NULL) {
-			net_pkt_print_frags(pkt);
-			res = net_recv_data(dev_data->iface, pkt);
-			if (res < 0) {
-				eth_stats_update_errors_rx(dev_data->iface);
-				LOG_ERR("Failed to enqueue frame "
-					"into RX queue: %d", res);
-				net_pkt_unref(pkt);
+		res = k_sem_take(&dev_data->rx_int_sem,
+			K_MSEC(CONFIG_ETH_STM32_CARRIER_CHECK_RX_IDLE_TIMEOUT_MS));
+		if (res == 0) {
+			/* semaphore taken, receive packets */
+			while ((pkt = eth_rx(dev)) != NULL) {
+				net_pkt_print_frags(pkt);
+				res = net_recv_data(dev_data->iface, pkt);
+				if (res < 0) {
+					eth_stats_update_errors_rx(dev_data->iface);
+					LOG_ERR("Failed to enqueue frame "
+						"into RX queue: %d", res);
+					net_pkt_unref(pkt);
+				}
+			}
+			/* received packet(s), update link status */
+			if (dev_data->link_up != true) {
+				dev_data->link_up = true;
+				net_eth_carrier_on(dev_data->iface);
+			}
```

Comment:
Doesn't it cause problems when net_recv_data() is called when the link was down, since the net_eth_carrier_on() is called after the data is received? 

Maybe move he test for link up to line 248 or so ?

Reply:
> Maybe move he test for link up to line 248 or so ?

Yep, it would be better to call `net_eth_carrier_on()` before feeding the data to upper stack.

Reply:
Thank you both for your feedback. The updated version now first checks the link status and, if necessary, calls `net_eth_carrier_on()` before receiving packets.